x/stake/keeper.go

```diff
@@ -33,319 +29,356 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
+//_________________________________________________________________________
+
+// get a single validator
+func (k Keeper) GetValidator(ctx sdk.Context, addr sdk.Address) (validator Validator, found bool) {
 	store := ctx.KVStore(k.storeKey)
-	b := store.Get(CounterKey)
+	b := store.Get(GetValidatorKey(addr))
 	if b == nil {
-		return 0
-	}
-	var counter int16
-	err := k.cdc.UnmarshalBinary(b, &counter)
-	if err != nil {
-		panic(err)
+		return validator, false
 	}
-	return counter
+	k.cdc.MustUnmarshalBinary(b, &validator)
+	return validator, true
 }
 
-// set the current in-block validator operation counter
-func (k Keeper) setCounter(ctx sdk.Context, counter int16) {
+// Get the set of all validators with no limits, used during genesis dump
+func (k Keeper) getAllValidators(ctx sdk.Context) (validators Validators) {
 	store := ctx.KVStore(k.storeKey)
-	bz, err := k.cdc.MarshalBinary(counter)
-	if err != nil {
-		panic(err)
-	}
-	store.Set(CounterKey, bz)
-}
+	iterator := store.SubspaceIterator(ValidatorsKey)
 
-//_________________________________________________________________________
-
-// get a single candidate
-func (k Keeper) GetCandidate(ctx sdk.Context, addr sdk.Address) (candidate Candidate, found bool) {
-	store := ctx.KVStore(k.storeKey)
-	b := store.Get(GetCandidateKey(addr))
-	if b == nil {
-		return candidate, false
-	}
-	err := k.cdc.UnmarshalBinary(b, &candidate)
-	if err != nil {
-		panic(err)
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() {
+			iterator.Close()
+			break
+		}
+		bz := iterator.Value()
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		validators = append(validators, validator)
+		iterator.Next()
 	}
-	return candidate, true
+	return validators[:i] // trim
 }
 
-// Get the set of all candidates, retrieve a maxRetrieve number of records
-func (k Keeper) GetCandidates(ctx sdk.Context, maxRetrieve int16) (candidates Candidates) {
+// Get the set of all validators, retrieve a maxRetrieve number of records
+func (k Keeper) GetValidators(ctx sdk.Context, maxRetrieve int16) (validators Validators) {
 	store := ctx.KVStore(k.storeKey)
-	iterator := store.Iterator(subspace(CandidatesKey))
+	iterator := store.SubspaceIterator(ValidatorsKey)
 
-	candidates = make([]Candidate, maxRetrieve)
+	validators = make([]Validator, maxRetrieve)
 	i := 0
 	for ; ; i++ {
 		if !iterator.Valid() || i > int(maxRetrieve-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
-		var candidate Candidate
-		err := k.cdc.UnmarshalBinary(bz, &candidate)
-		if err != nil {
-			panic(err)
-		}
-		candidates[i] = candidate
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		validators[i] = validator
 		iterator.Next()
 	}
-	return candidates[:i] // trim
+	return validators[:i] // trim
 }
 
-func (k Keeper) setCandidate(ctx sdk.Context, candidate Candidate) {
+func (k Keeper) setValidator(ctx sdk.Context, validator Validator) Validator {
```

Comment:
would be helpful to have some comments describing how this is going to go down - ie. what we have to do to update a validator - can also be reused in the spec

Reply:
could we simplify this function by instead having a `newValidator` method so we can remove all the checks for oldFound ? Might even be worth multiple functions that more clearly articulate in the method name how its being updated. eg. updateValidatorDescription and increase/decreaseValidatorBond. Could be more clear than putting them all under setValidator

Reply:
so I made an attempt to break up the `setValidator` (renamed `updateValidator`) and it became real ugly with lot's of code dupe. unfortunately there are lot's of more complex overlap scenarios aka (do `A` if `oldFound & something else` else do `B`)... because of this if we wanted to break it out more I think we'd be splitting out the code in a way that's sometimes slightly more confusing and generally a bit less maintainable - I'm going to keep thinking about this function, but making it better is not totally obvious atm.

Reply:
(addressed comment)